File: src/core/git-manager.ts
import { simpleGit, SimpleGit, CommitResult } from 'simple-git';
import * as path from 'path';
import { FileUtils } from '../utils/file-utils';
import { logger } from '../utils/logger';

export interface GitCommit {
  hash: string;
  message: string;
  author: string;
  date: Date;
}

export class GitManager {
  private git: SimpleGit;
  private repositoryPath: string;
  private initialized = false;

  constructor() {
    this.repositoryPath = '';
    this.git = simpleGit();
  }

  async initialize(registryPath: string): Promise<void> {
    this.repositoryPath = registryPath;
    this.git = simpleGit(this.repositoryPath);

    try {
      // Ensure registry directory exists
      await FileUtils.ensureDirectory(this.repositoryPath);

      // Check if git repository exists
      const isRepo = await this.isGitRepository();
      
      if (!isRepo) {
        logger.info('Initializing Git repository for registry');
        await this.git.init();
        
        // Create initial .gitkeep file
        const gitkeepPath = path.join(this.repositoryPath, '.gitkeep');
        await FileUtils.writeJson(gitkeepPath, {});
        
        // Create .gitignore for the registry repo
        const gitignorePath = path.join(this.repositoryPath, '.gitignore');
        const gitignoreContent = `# Runtime files
conversations/
backups/
*.log
*.tmp
`;
        await FileUtils.ensureDirectory(path.dirname(gitignorePath));
        await require('fs/promises').writeFile(gitignorePath, gitignoreContent, 'utf-8');
        
        // Configure git if not already configured
        await this.configureGit();
        
        // Initial commit
        await this.git.add('.gitkeep');
        await this.git.commit('Initial commit - Nox registry initialized');
        
        logger.info('Git repository initialized');
      } else {
        logger.info('Existing Git repository found');
      }

      this.initialized = true;

    } catch (error) {
      logger.error('Failed to initialize Git repository:', error);
      throw error;
    }
  }

  private async isGitRepository(): Promise<boolean> {
    try {
      await this.git.status();
      return true;
    } catch {
      return false;
    }
  }

  private async configureGit(): Promise<void> {
    try {
      // Check if user is already configured globally
      const globalConfig = await this.git.getConfig('user.name', 'global').catch(() => null);
      
      if (!globalConfig) {
        // Set default user for this repository
        await this.git.addConfig('user.name', 'Nox System', false);
        await this.git.addConfig('user.email', 'nox@localhost', false);
        logger.info('Git user configuration set for repository');
      }
    } catch (error) {
      logger.warn('Could not configure git user:', error);
    }
  }

  async commit(message: string, files?: string[]): Promise<string> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      // Add files to staging area
      if (files && files.length > 0) {
        await this.git.add(files);
      } else {
        // Add all changes
        await this.git.add('.');
      }

      // Check if there are changes to commit
      const status = await this.git.status();
      if (status.staged.length === 0) {
        logger.debug('No changes to commit');
        return await this.getCurrentCommit();
      }

      // Commit with timestamp and prefix
      const timestamp = new Date().toISOString();
      const commitMessage = `[nox] ${message} - ${timestamp}`;
      
      const result: CommitResult = await this.git.commit(commitMessage);
      logger.info(`Git commit created: ${result.commit} - ${message}`);
      
      return result.commit;

    } catch (error) {
      logger.error('Failed to create git commit:', error);
      throw error;
    }
  }

  async getCurrentCommit(): Promise<string> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const log = await this.git.log({ maxCount: 1 });
      return log.latest?.hash || 'no-commits';
    } catch (error) {
      logger.error('Failed to get current commit:', error);
      return 'unknown';
    }
  }

  async getCommitHistory(limit = 10): Promise<GitCommit[]> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const log = await this.git.log({ maxCount: limit });
      
      return log.all.map(commit => ({
        hash: commit.hash,
        message: commit.message,
        author: commit.author_name,
        date: new Date(commit.date)
      }));

    } catch (error) {
      logger.error('Failed to get commit history:', error);
      return [];
    }
  }

  async rollback(commitHash: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      logger.info(`Rolling back to commit: ${commitHash}`);
      
      // Create backup commit before rollback
      await this.commit('Backup before rollback');
      
      // Reset to specified commit
      await this.git.reset(['--hard', commitHash]);
      
      logger.info(`Rollback completed to commit: ${commitHash}`);

    } catch (error) {
      logger.error('Failed to rollback:', error);
      throw error;
    }
  }

  async rollbackToTime(hoursAgo: number): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const targetTime = new Date(Date.now() - (hoursAgo * 60 * 60 * 1000));
      const history = await this.getCommitHistory(100); // Get more history to find the right commit
      
      // Find the first commit before the target time
      const targetCommit = history.find(commit => commit.date <= targetTime);
      
      if (!targetCommit) {
        throw new Error(`No commit found ${hoursAgo} hours ago`);
      }

      logger.info(`Rolling back to commit from ${targetCommit.date.toISOString()}: ${targetCommit.hash}`);
      await this.rollback(targetCommit.hash);

    } catch (error) {
      logger.error('Failed to rollback to time:', error);
      throw error;
    }
  }

  async createBranch(branchName: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      await this.git.checkoutLocalBranch(branchName);
      logger.info(`Created and switched to branch: ${branchName}`);
    } catch (error) {
      logger.error(`Failed to create branch ${branchName}:`, error);
      throw error;
    }
  }

  async switchBranch(branchName: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      await this.git.checkout(branchName);
      logger.info(`Switched to branch: ${branchName}`);
    } catch (error) {
      logger.error(`Failed to switch to branch ${branchName}:`, error);
      throw error;
    }
  }

  async listBranches(): Promise<string[]> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const branches = await this.git.branchLocal();
      return branches.all;
    } catch (error) {
      logger.error('Failed to list branches:', error);
      return [];
    }
  }

  async getCurrentBranch(): Promise<string> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const branches = await this.git.branchLocal();
      return branches.current;
    } catch (error) {
      logger.error('Failed to get current branch:', error);
      return 'unknown';
    }
  }

  async mergeBranch(branchName: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      await this.git.merge([branchName]);
      logger.info(`Merged branch: ${branchName}`);
    } catch (error) {
      logger.error(`Failed to merge branch ${branchName}:`, error);
      throw error;
    }
  }

  async getStatus(): Promise<{ clean: boolean; staged: number; modified: number; untracked: number }> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const status = await this.git.status();
      
      return {
        clean: status.isClean(),
        staged: status.staged.length,
        modified: status.modified.length,
        untracked: status.not_added.length
      };
    } catch (error) {
      logger.error('Failed to get git status:', error);
      return { clean: false, staged: 0, modified: 0, untracked: 0 };
    }
  }

  async addRemote(name: string, url: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      await this.git.addRemote(name, url);
      logger.info(`Added remote: ${name} -> ${url}`);
    } catch (error) {
      logger.error(`Failed to add remote ${name}:`, error);
      throw error;
    }
  }

  async push(remote = 'origin', branch?: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const currentBranch = branch || await this.getCurrentBranch();
      await this.git.push(remote, currentBranch);
      logger.info(`Pushed to ${remote}/${currentBranch}`);
    } catch (error) {
      logger.error('Failed to push:', error);
      throw error;
    }
  }

  async pull(remote = 'origin', branch?: string): Promise<void> {
    if (!this.initialized) {
      throw new Error('Git manager not initialized');
    }

    try {
      const currentBranch = branch || await this.getCurrentBranch();
      await this.git.pull(remote, currentBranch);
      logger.info(`Pulled from ${remote}/${currentBranch}`);
    } catch (error) {
      logger.error('Failed to pull:', error);
      throw error;
    }
  }

  get isInitialized(): boolean {
    return this.initialized;
  }
}

File: src/core/git_manager.rs
//! Git manager module for the Nox agent ecosystem
//! 
//! This module handles Git operations for version control of the registry.

use anyhow::{anyhow, Result};
use log::{info, warn};
use std::path::Path;
use tokio::process::Command as TokioCommand;

/// Initialize a Git repository in the specified directory
pub async fn initialize_repo(repo_path: &Path) -> Result<()> {
    // Check if the directory is already a git repository
    if is_git_repo(repo_path).await? {
        info!("Git repository already initialized at {:?}", repo_path);
        return Ok(());
    }

    info!("Initializing git repository at {:?}", repo_path);

    // Initialize the repository
    let output = TokioCommand::new("git")
        .arg("init")
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to initialize git repository: {}", error_msg));
    }

    // Configure git user if not already configured
    if !is_git_user_configured().await? {
        configure_git_user().await?;
    }

    // Create .gitignore file
    create_gitignore(repo_path).await?;

    // Make initial commit
    commit_changes(repo_path, "Initial commit").await?;

    info!("Git repository initialized successfully");
    Ok(())
}

/// Check if a directory is a git repository
async fn is_git_repo(repo_path: &Path) -> Result<bool> {
    let output = TokioCommand::new("git")
        .args(&["rev-parse", "--is-inside-work-tree"])
        .current_dir(repo_path)
        .output()
        .await;

    match output {
        Ok(output) => Ok(output.status.success()),
        Err(_) => Ok(false),
    }
}

/// Check if git user is configured
async fn is_git_user_configured() -> Result<bool> {
    let name_output = TokioCommand::new("git")
        .args(&["config", "--get", "user.name"])
        .output()
        .await?;

    let email_output = TokioCommand::new("git")
        .args(&["config", "--get", "user.email"])
        .output()
        .await?;

    Ok(name_output.status.success() && email_output.status.success())
}

/// Configure git user for the repository
async fn configure_git_user() -> Result<()> {
    // Set default git user for the repository
    let name_output = TokioCommand::new("git")
        .args(&["config", "--global", "user.name", "Nox Agent"])
        .output()
        .await?;

    if !name_output.status.success() {
        let error_msg = String::from_utf8_lossy(&name_output.stderr);
        warn!("Failed to configure git user.name: {}", error_msg);
    }

    let email_output = TokioCommand::new("git")
        .args(&["config", "--global", "user.email", "nox-agent@example.com"])
        .output()
        .await?;

    if !email_output.status.success() {
        let error_msg = String::from_utf8_lossy(&email_output.stderr);
        warn!("Failed to configure git user.email: {}", error_msg);
    }

    Ok(())
}

/// Create a .gitignore file in the repository
async fn create_gitignore(repo_path: &Path) -> Result<()> {
    let gitignore_path = repo_path.join(".gitignore");

    // Default content for .gitignore
    let gitignore_content = r#"# Nox registry gitignore
# Ignore temporary files
*.tmp
*.temp
*.swp
*~

# Ignore logs
*.log

# Ignore OS specific files
.DS_Store
Thumbs.db
"#;

    // Write the .gitignore file
    std::fs::write(&gitignore_path, gitignore_content)?;

    info!("Created .gitignore file at {:?}", gitignore_path);
    Ok(())
}

/// Commit changes to the repository
pub async fn commit_changes(repo_path: &Path, message: &str) -> Result<()> {
    // Stage all changes
    let stage_output = TokioCommand::new("git")
        .args(&["add", "."])
        .current_dir(repo_path)
        .output()
        .await?;

    if !stage_output.status.success() {
        let error_msg = String::from_utf8_lossy(&stage_output.stderr);
        return Err(anyhow!("Failed to stage changes: {}", error_msg));
    }

    // Check if there are changes to commit
    let status_output = TokioCommand::new("git")
        .args(&["status", "--porcelain"])
        .current_dir(repo_path)
        .output()
        .await?;

    if status_output.stdout.is_empty() {
        info!("No changes to commit");
        return Ok(());
    }

    // Add timestamp to commit message for better tracking
    let timestamp = chrono::Utc::now().format("%Y-%m-%d %H:%M:%S UTC");
    let full_message = format!("{} [{}]", message, timestamp);

    // Commit changes
    let commit_output = TokioCommand::new("git")
        .args(&["commit", "-m", &full_message])
        .current_dir(repo_path)
        .output()
        .await?;

    if !commit_output.status.success() {
        let error_msg = String::from_utf8_lossy(&commit_output.stderr);

        // Check if it's a non-fatal error (like "nothing to commit")
        if error_msg.contains("nothing to commit") {
            info!("Nothing to commit: {}", error_msg.trim());
            return Ok(());
        }

        return Err(anyhow!("Failed to commit changes: {}", error_msg));
    }

    info!("Committed changes with message: {}", full_message);

    // Create a tag for significant changes if specified in the message
    if message.contains("[SIGNIFICANT]") {
        let tag_name = format!("v{}", timestamp.to_string().replace(" ", "-").replace(":", "-"));
        let tag_output = TokioCommand::new("git")
            .args(&["tag", &tag_name])
            .current_dir(repo_path)
            .output()
            .await?;

        if tag_output.status.success() {
            info!("Created tag: {}", tag_name);
        } else {
            let error_msg = String::from_utf8_lossy(&tag_output.stderr);
            warn!("Failed to create tag: {}", error_msg);
        }
    }

    Ok(())
}

/// Get the commit history for the repository
pub async fn get_commit_history(repo_path: &Path, limit: usize) -> Result<Vec<String>> {
    let output = TokioCommand::new("git")
        .args(&["log", "--pretty=format:%h %ad %s", "--date=short", &format!("-{}", limit)])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to get commit history: {}", error_msg));
    }

    let history = String::from_utf8_lossy(&output.stdout)
        .lines()
        .map(String::from)
        .collect();

    Ok(history)
}

/// Get the diff for a specific file
#[allow(dead_code)]
pub async fn get_file_diff(repo_path: &Path, file_path: &Path) -> Result<String> {
    let relative_path = file_path.strip_prefix(repo_path)?;

    let output = TokioCommand::new("git")
        .args(&["diff", "--", relative_path.to_str().unwrap()])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to get file diff: {}", error_msg));
    }

    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

/// Revert changes to a specific commit
pub async fn revert_to_commit(repo_path: &Path, commit_hash: &str) -> Result<()> {
    let output = TokioCommand::new("git")
        .args(&["reset", "--hard", commit_hash])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to revert to commit {}: {}", commit_hash, error_msg));
    }

    info!("Reverted to commit: {}", commit_hash);
    Ok(())
}

/// Create a new branch
pub async fn create_branch(repo_path: &Path, branch_name: &str) -> Result<()> {
    // Check if branch already exists
    let check_output = TokioCommand::new("git")
        .args(&["branch", "--list", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    let check_result = String::from_utf8_lossy(&check_output.stdout);
    if check_result.contains(branch_name) {
        return Err(anyhow!("Branch {} already exists", branch_name));
    }

    // Create the branch
    let output = TokioCommand::new("git")
        .args(&["branch", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to create branch {}: {}", branch_name, error_msg));
    }

    info!("Created branch: {}", branch_name);
    Ok(())
}

/// Switch to a branch
pub async fn switch_branch(repo_path: &Path, branch_name: &str) -> Result<()> {
    // Check if branch exists
    let check_output = TokioCommand::new("git")
        .args(&["branch", "--list", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    let check_result = String::from_utf8_lossy(&check_output.stdout);
    if !check_result.contains(branch_name) {
        return Err(anyhow!("Branch {} does not exist", branch_name));
    }

    // Switch to the branch
    let output = TokioCommand::new("git")
        .args(&["checkout", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to switch to branch {}: {}", branch_name, error_msg));
    }

    info!("Switched to branch: {}", branch_name);
    Ok(())
}

/// List all branches
pub async fn list_branches(repo_path: &Path) -> Result<Vec<String>> {
    let output = TokioCommand::new("git")
        .args(&["branch"])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to list branches: {}", error_msg));
    }

    let branches = String::from_utf8_lossy(&output.stdout)
        .lines()
        .map(|line| line.trim().trim_start_matches('*').trim().to_string())
        .filter(|line| !line.is_empty())
        .collect();

    Ok(branches)
}

/// Get current branch
pub async fn get_current_branch(repo_path: &Path) -> Result<String> {
    let output = TokioCommand::new("git")
        .args(&["branch", "--show-current"])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!("Failed to get current branch: {}", error_msg));
    }

    let branch = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if branch.is_empty() {
        return Err(anyhow!("Not on a branch (detached HEAD state)"));
    }

    Ok(branch)
}

/// Merge a branch into the current branch
pub async fn merge_branch(repo_path: &Path, branch_name: &str) -> Result<()> {
    // Check if branch exists
    let check_output = TokioCommand::new("git")
        .args(&["branch", "--list", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    let check_result = String::from_utf8_lossy(&check_output.stdout);
    if !check_result.contains(branch_name) {
        return Err(anyhow!("Branch {} does not exist", branch_name));
    }

    // Get current branch
    let current_branch = get_current_branch(repo_path).await?;

    // Merge the branch
    let output = TokioCommand::new("git")
        .args(&["merge", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);

        // Check if there are merge conflicts
        if error_msg.contains("CONFLICT") || error_msg.contains("Automatic merge failed") {
            // Abort the merge
            let _ = TokioCommand::new("git")
                .args(&["merge", "--abort"])
                .current_dir(repo_path)
                .output()
                .await?;

            return Err(anyhow!("Merge conflicts detected. Merge aborted. Please resolve conflicts manually."));
        }

        return Err(anyhow!("Failed to merge branch {} into {}: {}", branch_name, current_branch, error_msg));
    }

    info!("Merged branch {} into {}", branch_name, current_branch);
    Ok(())
}

/// Delete a branch
pub async fn delete_branch(repo_path: &Path, branch_name: &str, force: bool) -> Result<()> {
    // Check if branch exists
    let check_output = TokioCommand::new("git")
        .args(&["branch", "--list", branch_name])
        .current_dir(repo_path)
        .output()
        .await?;

    let check_result = String::from_utf8_lossy(&check_output.stdout);
    if !check_result.contains(branch_name) {
        return Err(anyhow!("Branch {} does not exist", branch_name));
    }

    // Get current branch
    let current_branch = get_current_branch(repo_path).await?;
    if current_branch == branch_name {
        return Err(anyhow!("Cannot delete the current branch. Switch to another branch first."));
    }

    // Delete the branch
    let mut args = vec!["branch"];
    if force {
        args.push("-D");
    } else {
        args.push("-d");
    }
    args.push(branch_name);

    let output = TokioCommand::new("git")
        .args(&args)
        .current_dir(repo_path)
        .output()
        .await?;

    if !output.status.success() {
        let error_msg = String::from_utf8_lossy(&output.stderr);

        // Check if branch is not fully merged
        if error_msg.contains("not fully merged") {
            return Err(anyhow!("Branch {} is not fully merged. Use --force to delete anyway.", branch_name));
        }

        return Err(anyhow!("Failed to delete branch {}: {}", branch_name, error_msg));
    }

    info!("Deleted branch: {}", branch_name);
    Ok(())
}
